tools/mkcomposefs.c

```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+{
+	pthread_mutex_lock( &mutex_NodeIterator );
+	cancelRequest = true;
+	pthread_mutex_unlock( &mutex_NodeIterator );
+}
+
+typedef int (*THREAD_PROCESS_PROC)(struct ParallelCopyInfo	*, void *);
+
+static int process_copy(struct ParallelCopyInfo	*item, void *digest_store_path)
+{
+	if (copy_file_with_dirs_if_needed(	item->parentPath, 
+										(const char*)digest_store_path, 
+										lcfs_node_get_payload(item->node), 
+										true) < 0)
+	{
+		return -1;
+	}
+	return 0;
+}
+
+static int process_compute(struct ParallelCopyInfo *item, void *)
+{
+	bool do_digest = lcfs_get_delayed_digest_calculation(item->node);
+	bool do_inline = lcfs_get_delayed_inline			(item->node);
+	bool by_digest = lcfs_get_delayed_payload			(item->node);
+
+	const char 			*fname 	= item->parentPath;
+	struct lcfs_node_s 	*ret 	= item->node;
+
+	// It is clear that at least digest or payload or inline is required.
+	cleanup_fd int fd =
+		openat(AT_FDCWD, fname, O_RDONLY | O_CLOEXEC);
+	if (fd < 0)
+		return fd;
+			
+	if (do_digest) 
+	{
+		int r = lcfs_node_set_fsverity_from_fd(ret, fd);
+		if (r < 0)
+			return r;
+
+		if (by_digest) 
+		{
+			const uint8_t *digest =
+				lcfs_node_get_fsverity_digest(ret);
+			char digest_path[LCFS_DIGEST_SIZE * 2 + 2];
+			lcfs_digest_to_path(digest, digest_path);
+			r = lcfs_node_set_payload(ret, digest_path);
+			if (r < 0)
+				return r;
+
+			/* We just computed digest to get the payoad path */
+			if(lcfs_get_reset_digest(ret))
+				ret->digest_set = false;
+		}
+	}
+	if (do_inline) {
+		lseek(fd, 0, SEEK_SET);
+		uint8_t buf[LCFS_BUILD_INLINE_FILE_SIZE_LIMIT];
+		struct stat sb;
+		int r = fstatat(AT_FDCWD, fname, &sb, AT_SYMLINK_NOFOLLOW);
+		if (r < 0)
+			return r;
+
+		r = lcfs_read_content(fd, sb.st_size, buf);
+		if (r < 0)
+			return r;
+
+		r = lcfs_node_set_content(ret, buf, sb.st_size);
+		if (r < 0)
+			return r;
+	}
+
+	return 0;
+}
+
+struct ThreadData
+{
+	THREAD_PROCESS_PROC proc;
+	void 				*data;
+};
+
+static void* thread_proc(void *data)
+{
+	struct ThreadData *info = (struct ThreadData*)data;
+
+	while( true )
+	{
+		struct ParallelCopyInfo	*item = get_next_node();
+
+		if(!item) 		return 0;
+		if(!item->node)	return 0; // The last item has an empty node. TODO
+
+		if (info->proc(item, info->data) != 0)
+		{
+			request_cancel();
+			return 0;
+		}
+	}
+	return 0;
+}
+
+static bool execute_in_threads(const int nrThreads, 
+					struct ParallelCopyInfo *item, 
+					THREAD_PROCESS_PROC proc, 
+					void *data )
+{
+	nextNode 		= item;
+	cancelRequest 	= false;
+	
+	struct ThreadData threadInfo;
+	threadInfo.data = data;
+	threadInfo.proc = proc;
+
+	log_message_with_time( "executeInThreads: begin" );	
+	fflush(stdout); 
+
+	pthread_t * threads = calloc(nrThreads, 
+							sizeof(pthread_t));
+	for( int i = 0; i< nrThreads; i++) 
+		pthread_create(&threads[i], 
+					NULL, 
+					thread_proc, 
+					&threadInfo);
+
+	for( int i = 0; i< nrThreads; i++) 
+		pthread_join(threads[i], NULL);
+
```

Comment:
I think this should create n_threads - 1 thread, and then call thread_proc on the main thread as well. Then we can efficiently handle the n_thread == 1 case without creating any threads.

Reply:
I liked this proposal. I changed it.